add support for using nnpack when installed via conda

killeent opened the pull request:
Pairs with https://github.com/pytorch/builder/pull/14. This basically searches for the appropriate `NNPACK` libraries and headers in the Conda installation environment, and builds with them. We move the headers to `tmp_install/include` so that we don't pass the conda include dir to the build, which seems like a bad idea.

cc @ezyang

killeent committed:
add support for using nnpack when installed via conda

ezyang commented on setup.py:
```diff
@@ -507,9 +513,21 @@ def check_file(f):
+    else:
+        # Conda Defined
+        main_link_args.extend(NNPACK_LIB_PATHS)
+        import pdb; pdb.set_trace()
```
Oops.

ezyang commented on setup.py:
```diff
@@ -230,7 +232,11 @@ def run(self):
+            if NNPACK_IS_USER_DEFINED:
+                nnpack_dir = NNPACK_LIB_DIR
+            else:
+                nnpack_dir = NNPACK_LIB_PATHS[0]
```
Code like this always makes me nervous (what if it's empty? What about the rest of it?) but I guess you had a good reason for doing it this way...

apaszke replied:
This should also be inside the tool. Why does setup.py need to know what’s the difference between NNPACK_LIB_DIR and NNPACK_LIB_PATHS. to me it seems that they’re pretty much the same thing

killeent replied:
@ezyang its sort of an invariant that to hit this code path this list should not be empty.

apaszke commented on setup.py:
```diff
@@ -507,9 +513,21 @@ def check_file(f):
+        main_libraries += ['nnpack', 'pthreadpool']
+        include_dirs.extend(NNPACK_INCLUDE_DIRS)
+        library_dirs.append(NNPACK_LIB_DIR)
+    else:
```
It seems to me that this logic should be in the nnpack tool. setup.py should be a simple scripts that only gathers information and adds it to the extension

killeent replied:
I disagree. I need this code logic to be in `setup.py` because it sets up where we install temporary files/headers etc. which `setup_helpers` do not know about. Otherwise, if we change `tmp_install` to `temp_install` or something, then NNPACK builds break and you have to look and manually change a second file to fix this. I'd rather have this patchwork duplication in two places.

apaszke replied:
Oh I didn't notice that you copy the files to `tmp_install`! But why? Just add the dir where NNPACK headers are. We don't do this for any other dependency

killeent replied:
@apaszke - The headers are in the user's conda include directory. I discussed this with @ezyang and he suggested not to add this to `include_dirs` as then we are picking up all the other headers as well, which I guess can be problematic.

apaszke replied:
Compilers look for headers in a well defined order. Everything should be fine as long as you add this dir at the end of include dirs.

killeent commented:
@apaszke if you'd like, I can move to a model where the "user-defined" `NNPACK` case also passes a list of paths to the libraries themselves, which would eliminate the need for separate `LIB_DIR` vs. `LIB_PATHS`.

killeent committed:
unify nnpack discovery between conda and user

killeent commented:
Okay, I think I addressed all of the comments. Now both user defined and conda-installed NNPACK operate the same way, other than discovering where to find them.

prigoyal commented:
FYI, nnpack has full CMAKE build support as well now https://github.com/Maratyszcza/NNPACK/commit/dd0e4637190734536cd947cd45bf09b2af68e26d

killeent commented:
@prigoyal yep I already tested it out. I will update https://github.com/pytorch/builder/blob/master/conda/nnpack/build.sh when I can. Waiting to be able to test things on the actual jenkins boxes.
